src/gc/gc.cpp

```diff
@@ -28722,13 +28862,18 @@ inline void
 gc_heap::mark_through_cards_helper (uint8_t** poo, size_t& n_gen,
                                     size_t& cg_pointers_found,
                                     card_fn fn, uint8_t* nhigh,
-                                    uint8_t* next_boundary)
+                                    uint8_t* next_boundary,
+                                    gc_heap* hpt)
 {
+#if defined(FEATURE_CARD_MARKING_STEALING) && defined(MULTIPLE_HEAPS)
+    int thread = hpt->heap_number;
+#else
     THREAD_FROM_HEAP;
+#endif
     if ((gc_low <= *poo) && (gc_high > *poo))
     {
         n_gen++;
-        call_fn(fn) (poo THREAD_NUMBER_ARG);
+        call_fn(hpt,fn) (poo THREAD_NUMBER_ARG);
```

Comment:
>call_fn(hpt,fn) (poo THREAD_NUMBER_ARG); [](start = 8, length = 40)

for mark phase, this will call mark_object_simple with hpt's gc_low/gc_high instead of with the gc_low/gc_high of the heap it's marking. so in gc_mark now the order is different which heap's gc_low/gc_high it's comparing with first

Reply:
Agreed, but is this a correctness problem? It's probably a small perf problem, I wonder how we can fix it...

Reply:
right, not a correctness problem. one could construct a scenario where the perf matters - for example if you allocated a bunch of static objects on your worker threads in a fairly balanced fashion which means they'd be on their cores' respective heaps, then these objects refer to temp objects created in the same fashion which means gc_mark will find the gc_low/gc_high on the first try. now, let's say some of the work we do before marking through cards happen to be very unbalanced so a lot of stealing will incur during card marking so now gc_mark will have to go to other heaps. 

currently I don't have better ideas than having mark_object_simple also take a gc_heap* arg (which is of course also cost).

Reply:
I would think that it would still be a net win - if thread 1 is fast and does part of thread 2's work, then thread 2 will finish faster even if thread 1 is less effective at it than thread 2 would be. We are essentially using time where thread 1 would be idle otherwise.

I don't better ideas for a fix to this issue either.

Reply:
Hi. As this features just landed I'm very excited about it but would also like to understand the changes.

Is it guaranteed that the following scenario won't happen?
Thread 1 finishes fast and starts working on thread 2's work.
Thread 2 had *almost* finished by the time thread 1 started to do its work.
Because thread 2 is now idle it starts doing thread 3's work, and so on...
Ultimately resulting in every thread working on someone else's work.

Thanks so much for all the new great features in .NET 5!

Reply:
it's totally fine if T2 finishes the end of T1's work and T3 finishes the end of T2's work. threads with a lot of work left will still be working on their respective heaps. only threads that are done with their own heaps would go steal other threads' work. does this make sense?